packages/admin/cms-admin/src/common/MasterMenuRoutes.tsx

```diff
@@ -7,24 +7,36 @@ import { MasterMenuData, MasterMenuItem } from "./MasterMenu";
 
 export function useRoutePropsFromMasterMenuData(items: MasterMenuData): RouteProps[] {
     const isAllowed = useUserPermissionCheck();
-    const checkPermission = (item: MasterMenuItem): boolean => !item.requiredPermission || isAllowed(item.requiredPermission);
+    const checkPermission = (item: MasterMenuItem, parent?: MasterMenuItem): boolean => {
+        if (item.requiredPermission) {
+            return isAllowed(item.requiredPermission);
+        } else if (parent?.requiredPermission) {
+            return isAllowed(parent.requiredPermission);
+        } else {
+            return true;
+        }
+    };
 
-    const flat = (routes: RouteProps[], item: MasterMenuItem & { icon?: ReactNode }): RouteProps[] => {
+    const flat = (
+        routes: RouteProps[],
+        item: MasterMenuItem & { icon?: ReactNode },
+        parent?: MasterMenuItem & { icon?: ReactNode },
+    ): RouteProps[] => {
         if (item.type === "externalLink") {
             return routes;
         }
         if (item.type === "group") {
-            return routes.concat(item.items.reduce(flat, []));
+            return routes.concat(item.items.reduce((routes, child) => flat(routes, child, item), [] as RouteProps[]));
         }
-        if (item.route && checkPermission(item)) {
+        if (item.route && checkPermission(item, parent)) {
             routes.push(item.route);
         }
         if (item.type === "collapsible" && !!item.items?.length) {
-            routes.concat((item.items as Array<MasterMenuItem & { icon?: ReactNode }>).reduce(flat, routes));
+            routes.concat((item.items as Array<MasterMenuItem & { icon?: ReactNode }>).reduce((routes, child) => flat(routes, child, item), routes));
         }
         return routes;
     };
-    return items.reduce(flat, []);
+    return items.reduce((routes, item) => flat(routes, item), [] as RouteProps[]);
```

Comment:
Using reduce makes this complicate to read. I may change this to a for-loop or flatMap in a follow-up PR.